qucs/qucs-lib/qucslib.1
.TH QucsLib "1" "July 2005" "Debian/GNU Linux" "User Commands"
.SH NAME
QucsLib \- A component library manager for Qucs.
.SH SYNOPSIS
.B qucslib
[\fIOPTION\fR] ...
.SH DESCRIPTION

\fBQucs\fR is an integrated circuit simulator which means you are able
to setup a circuit with a graphical user interface (GUI) and simulate
the large-signal, small-signal and noise behaviour of the circuit.
After that simulation has finished you can view the simulation results
on a presentation page or window.

The software aims to support all kinds of circuit simulation types,
e.g. DC, AC, S-parameter, harmonic balance analysis, noise analysis,
etc.

\fBQucsLib\fR is a program to manage Qucs component libraries. On the
left side of the application window the available libraries can be
browsed to find the wanted component.  By clicking on the component
name its description can be seen on the right side. The selected
component is transported to the Qucs application by clicking on the
button "Copy to Clipboard".  Being back in the schematic window the
component can be inserted by pressing CTRL-V (paste from clipboard).

.SH AVAILABILITY
The latest version of Qucs can always be obtained from
\fBwww.sourceforge.net\fR or \fBwww.freshmeat.net\fR
.SH "REPORTING BUGS"
Known bugs are documented within the BUGS file.  Report bugs to
<qucs-bugs@lists.sourceforge.net>.
.SH COPYRIGHT
Copyright \(co 2005 Michael Margraf <michael.margraf@alumni.tu-berlin.de>
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
Written by Michael Margraf <michael.margraf@alumni.tu-berlin.de> and
Stefan Jahn <stefan@lkcc.org>.


NEWS
--
-- NEWS
--
-- Copyright (C) 2003, 2004, 2005 Stefan Jahn <stefan@lkcc.org>
--
-- This is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2, or (at your option)
-- any later version.
--
-- This software is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this package; see the file COPYING.  If not, write to
-- the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
-- Boston, MA 02110-1301, USA.
--

This file lists user visible changes that were made between releases.  For
more verbose descriptions of these and internal changes see the ChangeLog
files.

Version 0.0.7
-------------

 * text painting rotations in subcircuit symbols
 * renamed Qucs, new name is "Quite Universal Circuit Simulator"
 * several small improvements and bug fixes
 * Latex-like mnemonics (e.g. \tau) in graphic texts and diagram labels
 * library manager including transistor, diode, substrate, etc. libraries

Version 0.0.6
-------------

 * finished SPICE component
 * transmission line calculator
 * filter synthesis application
 * new diagrams: 3-dimensional cartesian, locus curve
 * many small improvements
 * edit component properties directly on the schematic
 * translations into Swedish
 * new components: coplanar open, short, gap and step
 * nicer icons for simulation components

Version 0.0.5
-------------

 * internal help system now also in German available
 * translations into Hebrew, French, Japanese and Hungarian
 * some new components and diagrams
 * some new graph styles (stars, circles, ...)
 * manual axes scale for diagrams
 * many minor improvements and bug fixes
 * improved simulation component dialog
 * scroll bar for tabular diagrams
 * menu on right mouse button

Version 0.0.4
-------------

 * per-language internal help system implemented
 * translations into portuguese and spanish
 * some new paintings and components
 * subcircuits with symbol painted by the user
 * improved the zooming (looks nicer now)
 * progress bar during simulation
 * some small improvements and bug fixes

Version 0.0.3
-------------

 * avoided working directory changes during program execution
 * translations into Polish and Romanian
 * some new components (e.g. transistors)
 * some new operations (e.g. aligning)
 * undo/redo function
 * improved subcircuit support
 * resize operation for paintings
 * logarithmical scale for rectangle diagrams
 * change style of graphs (dashed, dotted, ...)
 * qucs-edit and qucs-help as external programs
 * application settings dialog
 * many small improvements and bug fixes

Version 0.0.2
-------------

 * translations into german and italian
 * modified build system, no more kdevelop
 * a (very modest) help system
 * deleting and rename of documents
 * some new components and paintings
 * runs more stable
 * mouse cursor symbols show active action
 * document area more flexible for resize and zoom
 * equation support
 * strongly improved handling with wire labels
 * diagrams more beautiful
 * diagrams can show independent and more-dimensional graphs
 * markers can be set on graphs

Version 0.0.1
-------------

 * first available version


qucs-lib/qucslib.cpp
/***************************************************************************
                               qucslib.cpp
                              -------------
    begin                : Sat May 28 2005
    copyright            : (C) 2005 by Michael Margraf
    email                : michael.margraf@alumni.tu-berlin.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <qmenubar.h>
#include <qaction.h>
#include <qpopupmenu.h>
#include <qcombobox.h>
#include <qclipboard.h>
#include <qapplication.h>
#include <qlayout.h>
#include <qhbox.h>
#include <qvgroupbox.h>
#include <qpushbutton.h>
#include <qmessagebox.h>
#include <qtextedit.h>
#include <qlistbox.h>
#include <qdir.h>
#include <qregexp.h>

#include "qucslib.h"
#include "displaydialog.h"
#include "symbolwidget.h"
#include "searchdialog.h"


/* Constructor setups the GUI. */
QucsLib::QucsLib()
{
  // set application icon
  setIcon (QPixmap(QucsSettings.BitmapDir + "big.qucs.xpm"));
  setCaption("Qucs Library Tool " PACKAGE_VERSION);

  QMenuBar * menuBar = new QMenuBar (this);

  // create file menu
  QPopupMenu * fileMenu = new QPopupMenu ();
  QAction * fileQuit =
    new QAction (tr("Quit"), tr("&Quit"), CTRL+Key_Q, this);
  fileQuit->addTo (fileMenu);
  connect(fileQuit, SIGNAL(activated()), SLOT(slotQuit()));

  // create help menu
  QPopupMenu * helpMenu = new QPopupMenu ();
  QAction * helpHelp =
    new QAction (tr("Help"), tr("&Help"), Key_F1, this);
  helpHelp->addTo (helpMenu);
  connect(helpHelp, SIGNAL(activated()), SLOT(slotHelp()));
  QAction * helpAbout =
    new QAction (tr("About"), tr("About"), 0, helpMenu);
  helpAbout->addTo (helpMenu);
  connect(helpAbout, SIGNAL(activated()), SLOT(slotAbout()));

  // setup menu bar
  menuBar->insertItem (tr("&File"), fileMenu);
  menuBar->insertSeparator ();
  menuBar->insertItem (tr("&Help"), helpMenu);

  // main box
  QVBoxLayout * all = new QVBoxLayout (this);
  all->setSpacing (0);
  all->setMargin (0);

  // reserve space for menubar
  QWidget * Space = new QWidget (this);
  Space->setFixedSize(5, menuBar->height() + 2);
  all->addWidget (Space);

  // main layout
  QHBox * h = new QHBox (this);
  h->setSpacing (5);
  h->setMargin (3);
  all->addWidget (h);

  // library and component choice
  QVGroupBox * LibGroup = new QVGroupBox (tr("Component Selection"), h);
  Library = new QComboBox (LibGroup);
  connect(Library, SIGNAL(activated(int)), SLOT(slotSelectLibrary(int)));
  CompList = new QListBox(LibGroup);
  connect(CompList, SIGNAL(highlighted(QListBoxItem*)),
	SLOT(slotShowComponent(QListBoxItem*)));

  QHBox * h1 = new QHBox (LibGroup);
  QPushButton * SearchButton = new QPushButton (tr("Search..."), h1);
  connect(SearchButton, SIGNAL(clicked()), SLOT(slotSearchComponent()));
  h1->setStretchFactor(new QWidget(h1), 5); // stretchable placeholder


  // component display
  QVGroupBox *CompGroup = new QVGroupBox (tr("Component"), h);
  CompDescr = new QTextEdit(CompGroup);
  CompDescr->setTextFormat(Qt::PlainText);
  CompDescr->setReadOnly(true);
  CompDescr->setWordWrap(QTextEdit::NoWrap);

  Symbol = new SymbolWidget (CompGroup);

  QHBox * h2 = new QHBox (CompGroup);
  QPushButton * CopyButton = new QPushButton (tr("Copy to clipboard"), h2);
  connect(CopyButton, SIGNAL(clicked()), SLOT(slotCopyToClipBoard()));
  QPushButton * ShowButton = new QPushButton (tr("Show Model"), h2);
  connect(ShowButton, SIGNAL(clicked()), SLOT(slotShowModel()));

  // ......................................................
  // Put all available libraries into ComboBox.
  QDir LibDir(QucsSettings.LibDir);
  QStringList LibFiles = LibDir.entryList("*.lib", QDir::Files, QDir::Name);

  QStringList::iterator it;
  for(it = LibFiles.begin(); it != LibFiles.end(); it++)
    Library->insertItem((*it).left((*it).length()-4));

  slotSelectLibrary(0);
}

/* Destructor destroys the application. */
QucsLib::~QucsLib()
{
}

void QucsLib::slotAbout()
{
  QMessageBox::about(this, tr("About..."),
    "QucsLib " PACKAGE_VERSION "\n"+
    tr("Library Manager for Qucs\n")+
    tr("Copyright (C) 2005 by Michael Margraf\n")+
    "\nThis is free software; see the source for copying conditions."
    "\nThere is NO warranty; not even for MERCHANTABILITY or "
    "\nFITNESS FOR A PARTICULAR PURPOSE.");
}

void QucsLib::slotQuit()
{
  int tmp;
  tmp = x();		// call size and position function in order to ...
  tmp = y();		// ... set them correctly before closing the ...
  tmp = width();	// dialog !!!  Otherwise the frame of the window ...
  tmp = height();	// will not be recognized (a X11 problem).

  accept();
}

void QucsLib::closeEvent(QCloseEvent *Event)
{
  int tmp;
  tmp = x();		// call size and position function in order to ...
  tmp = y();		// ... set them correctly before closing the ...
  tmp = width();	// dialog !!!  Otherwise the frame of the window ...
  tmp = height();	// will not be recognized (a X11 problem).

  Event->accept();
}

void QucsLib::slotHelp()
{
  DisplayDialog *d = new DisplayDialog(this);
  d->setCaption("QucsLib Help");
  d->resize(250, 300);
  d->Text->setText(
     tr("QucsLib is a program to manage Qucs component libraries. "
	"On the left side of the application window the available "
	"libraries can be browsed to find the wanted component. "
	"By clicking on the component name its description can be "
	"seen on the right side. The selected component is "
	"transported to the Qucs application by clicking on the "
	"button \"Copy to Clipboard\". Being back in the schematic "
	"window the component can be inserted by pressing CTRL-V "
	" (paste from clipboard)."));
  d->show();
}

void QucsLib::slotCopyToClipBoard()
{
  QString s = "<Qucs Schematic " PACKAGE_VERSION ">\n";
  s += "<Components>\n  ";
  if(ModelString.contains('\n') < 2)
    s += ModelString;
  else
    s += "<Lib " + Symbol->Prefix + " 1 0 0 " +
         QString::number(Symbol->Text_x) + " " +
         QString::number(Symbol->Text_y) + " 0 0 \"" +
         Symbol->LibraryName + "\" 0 \"" + Symbol->ComponentName + "\" 0>\n";
  s += "</Components>\n";

  // put resulting schematic into clipboard
  QClipboard *cb = QApplication::clipboard();
  cb->setText(s);
}

void QucsLib::slotShowModel()
{
  DisplayDialog *d = new DisplayDialog(this);
  d->setCaption("Model");
  d->resize(500, 150);
  d->Text->setText(ModelString);
  d->Text->setWordWrap(QTextEdit::NoWrap);
  d->show();
}

void QucsLib::slotSelectLibrary(int Index)
{
  int Start, End, NameStart, NameEnd;
  End = Library->count()-1;
  if(Library->text(End) == tr("Search result"))
    if(Index < End)
      Library->removeItem(End); // if search result still there -> remove it
    else  return;


  CompList->clear();
  LibraryComps.clear();

  QFile file(QucsSettings.LibDir + Library->text(Index) + ".lib");
  if(!file.open(IO_ReadOnly)) {
    QMessageBox::critical(this, tr("Error"),
        tr("Cannot open \"%1\".").arg(
           QucsSettings.LibDir + Library->text(Index) + ".lib"));
    return;
  }

  QTextStream ReadWhole(&file);
  QString LibraryString = ReadWhole.read();
  file.close();

  Start = LibraryString.find("<Qucs Library ");
  if(Start < 0) {
    QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
    return;
  }
  End = LibraryString.find('>', Start);
  if(End < 0) {
    QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
    return;
  }
  QString LibName = LibraryString.mid(Start, End-Start).section('"', 1, 1);

  while((Start=LibraryString.find("\n<Component ", Start)) > 0) {
    Start++;
    NameStart = Start + 11;
    NameEnd = LibraryString.find('>', NameStart);
    if(NameEnd < 0)  continue;

    End = LibraryString.find("\n</Component>", NameEnd);
    if(End < 0)  continue;
    End += 13;

    CompList->insertItem(LibraryString.mid(NameStart, NameEnd-NameStart));
    LibraryComps.append(LibName+'\n'+LibraryString.mid(Start, End-Start));
    Start = End;
  }

  CompList->setSelected(0, true);  // select first item
}

void QucsLib::slotSearchComponent()
{
  SearchDialog *d = new SearchDialog(this);
  if(d->exec() != QDialog::Accepted)
    QMessageBox::information(this, tr("Result"),
                             tr("No appropriate component found."));
}

void QucsLib::slotShowComponent(QListBoxItem *Item)
{
  if(!Item) return;

  QStringList::Iterator CompString = LibraryComps.at(CompList->index(Item));
  CompDescr->setText("Name: " + Item->text());
  CompDescr->append("Library: " + (*CompString).section('\n', 0, 0));
  CompDescr->append("----------------------------");

  int Start, End;
  Start = (*CompString).find("<Description>");
  if(Start > 0) {
    Start += 13;
    End = (*CompString).find("</Description>", Start);
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }
    CompDescr->append(
      (*CompString).mid(Start, End-Start).replace(QRegExp("\\n\\x20+"), "\n").remove(0, 1));
  }


  Start = (*CompString).find("<Model>");
  if(Start > 0) {
    Start += 7;
    End = (*CompString).find("</Model>", Start);
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }
    ModelString =
      (*CompString).mid(Start, End-Start).replace(QRegExp("\\n\\x20+"), "\n").remove(0, 1);

    if(ModelString.contains('\n') < 2)
      Symbol->createSymbol(ModelString,
                           (*CompString).section('\n', 0, 0), Item->text());
  }


  Start = (*CompString).find("<Symbol>");
  if(Start > 0) {
    Start += 8;
    End = (*CompString).find("</Symbol>", Start);
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }
    Symbol->setSymbol((*CompString).mid(Start, End-Start),
                      (*CompString).section('\n', 0, 0), Item->text());
  }
}
